### DISCUSSION TITLE
Question on FluxRegion vs. DFT field (Python)

### DISCUSSION THREAD

**Discussion by theogdoctorg** [NanoComp/meep#2732] [Q&A] [answered]
Hello, I am trying to understand the difference between a FluxRegion and the DFT_Fields commands in the Python interface.

My motivation for this is to reproduce the values from a FluxRegion using the DFT_Fields post-simulation, so that I can separate out in the contribution of one electric field component from the other in the calculation of the Poynting flux, and thus the different power contributions of the two (linear) polarizations in only a single simulation run. If there's another way to do this (other than running two simulations), please do let me know!

Following the example in the "Computing Quantities Defined by Integrals of Field-Dependent Functions Over Grid Regions" section of the Python User Interface documents with the header "  # x-directed Poynting flux through monitor from frequency-domain fields", here's the abbreviated code:
```
    # refl plane loc will be at half of the cell height (in air)
    refl_fr = mp.FluxRegion(center=mp.Vector3(z = -z_height/2))
    refl = sim.add_flux(fcen, df, nfreq, refl_fr)
    
    # add dft flux and check if it matches "FluxRegion"
    dft_fields = sim.add_dft_fields([mp.Ex, mp.Ey, mp.Hx,mp.Hy], fcen, df, nfreq,
                       center = mp.Vector3(z = -z_height/2),
                       size = mp.Vector3(sim_xy_size,sim_xy_size,0))

    sim.run(until_after_sources = mp.stop_when_fields_decayed(
        50, mp.Ex, mp.Vector3(z = -z_height + dpml), 1e-9))
    
    # show fluxes from flux calcuation
    sim.display_fluxes(refl)
    # attempt "manual" calculation of 
    # the below line works equally well with "refl" and "dft_fields"
    (Ey,Ex,Hy,Hx) = [[sim.get_dft_array(refl,c,n) for n in range(nfreq)] for c in [mp.Ey, mp.Ex, mp.Hy, mp.Hx]]
    flux_density   = np.real( np.conj(Ex)*Hy - np.conj(Ey)*Hx )    # array
    manual_flux = freqs*flux_density          # freqs is array of freq values
```

The first few frequency results from sim.display_fluxes(refl)
flux1:, 1.25, 3.679463323024969e-06
flux1:, 1.2755102040816326, 8.435252483188409e-06
flux1:, 1.3010204081632653, 1.8695315730562005e-05

From the "manual_flux" variable calculated using refl (the FluxRegion):
7.18475e-08
1.68105e-07
3.8011e-07

From the "manual_flux" variable calculated using dft_fields:
7.18475e-08
1.68105e-07
3.8011e-07

Since the "manual" values are ~2 orders of magnitude off from the display_fluxes(...) values, I suspect that they represent the values of the Poynting vector at the end of the simulation instead of having been accumulated over a few hundred timesteps. Thus, I gather that somehow the FluxRegion is able to accumulate the result of a field function (Poynting calculation), while the DFT_field does not. Is this correct? Otherwise, is there something incorrect in my "manual" calculation?

**Comment by smartalecH** [comment, marked as answer]
You have a lot of questions, I'll just focus on the root of what you're trying to do:

> and thus the different power contributions of the two (linear) polarizations in only a single simulation run

Typically we use mode monitors for this (although you should be able to make this work with dft monitors too if you're careful). 

You can overlap with as many modes (or polarization states) as you want with a single run. If you do an infinite number, then the sum of their powers converges to the poynting flux.

**Reply by theogdoctorg** [reply]
Hi Alec, thanks for the reply! If I understand the mode decomposition correctly from [here](https://meep.readthedocs.io/en/latest/Mode_Decomposition/) it uses MPB and can only be used with lossless dielectric structures with wavelength-independent epsilon, which is not useful in my case. In the normal FDTD simulation, I expect that by starting to sources I could separate out the two polarizations in the accumulated fluxes. Currently, the simulation with a source in the Z-direction with component mp.Ex results in 0 values in the accumulated fluxes for mp.Ey and mp.Hx, as expected, and similarly mp.Ey -> zeros for mp.Ex and mp.Hy. So it seems like a simulation with two sources should produce the data needed to calculate Sz for the polariztions independently.

I tried to limit myself to two questions originally :) if you wouldn't mind taking another look, I'd appreciate it!
